test/common/config/subscription_factory_test.cc

```diff
@@ -60,21 +60,131 @@ TEST_F(SubscriptionFactoryTest, NoConfigSpecifier) {
+      "foo");
+  cluster_map.emplace("foo", cluster);
+
+  envoy::api::v2::core::GrpcService expected_grpc_service;
+  expected_grpc_service.mutable_envoy_grpc()->set_cluster_name("foo");
+
+  EXPECT_CALL(cm_, clusters()).WillOnce(Return(cluster_map));
+  EXPECT_CALL(cm_, grpcAsyncClientManager()).WillOnce(ReturnRef(cm_.async_client_manager_));
+  EXPECT_CALL(cm_.async_client_manager_,
+              factoryForGrpcService(ProtoEq(expected_grpc_service), _, _))
+      .WillOnce(Invoke([](const envoy::api::v2::core::GrpcService&, Stats::Scope&, bool) {
+        auto async_client_factory = std::make_unique<Grpc::MockAsyncClientFactory>();
+        EXPECT_CALL(*async_client_factory, create()).WillOnce(Invoke([] {
+          return std::make_unique<NiceMock<Grpc::MockAsyncClient>>();
+        }));
+        return async_client_factory;
+      }));
+  EXPECT_CALL(*cluster.info_, addedViaApi()).WillOnce(Return(false));
+  EXPECT_CALL(*cluster.info_, type()).WillOnce(Return(envoy::api::v2::Cluster::STATIC));
+  EXPECT_CALL(dispatcher_, createTimer_(_));
+
+  subscriptionFromConfigSource(config);
+}
+
+TEST_F(SubscriptionFactoryTest, RestClusterMultiton) {
+  envoy::api::v2::core::ConfigSource config;
+  Upstream::ClusterManager::ClusterInfoMap cluster_map;
+  NiceMock<Upstream::MockCluster> cluster;
+
+  config.mutable_api_config_source()->set_api_type(envoy::api::v2::core::ApiConfigSource::REST);
+
   config.mutable_api_config_source()->add_cluster_names("foo");
+  cluster_map.emplace("foo", cluster);
+
   config.mutable_api_config_source()->add_cluster_names("bar");
-  EXPECT_CALL(cm_, clusters());
+  cluster_map.emplace("bar", cluster);
+
+  EXPECT_CALL(cm_, clusters()).WillRepeatedly(Return(cluster_map));
+  EXPECT_CALL(*cluster.info_, addedViaApi()).WillRepeatedly(Return(false));
+  EXPECT_CALL(*cluster.info_, type()).WillRepeatedly(Return(envoy::api::v2::Cluster::STATIC));
   EXPECT_THROW_WITH_MESSAGE(
       subscriptionFromConfigSource(config), EnvoyException,
       "envoy::api::v2::core::ConfigSource must have a singleton cluster name specified");
 }
 
+TEST_F(SubscriptionFactoryTest, GrpcClusterMultiton) {
+  envoy::api::v2::core::ConfigSource config;
+  Upstream::ClusterManager::ClusterInfoMap cluster_map;
+  NiceMock<Upstream::MockCluster> cluster;
+
+  config.mutable_api_config_source()->set_api_type(envoy::api::v2::core::ApiConfigSource::GRPC);
+
+  config.mutable_api_config_source()->add_grpc_services()->mutable_envoy_grpc()->set_cluster_name(
+      "foo");
+  cluster_map.emplace("foo", cluster);
+  envoy::api::v2::core::GrpcService expected_grpc_service_foo;
+  expected_grpc_service_foo.mutable_envoy_grpc()->set_cluster_name("foo");
+
+  config.mutable_api_config_source()->add_grpc_services()->mutable_envoy_grpc()->set_cluster_name(
+      "bar");
+  cluster_map.emplace("bar", cluster);
+  envoy::api::v2::core::GrpcService expected_grpc_service_bar;
+  expected_grpc_service_bar.mutable_envoy_grpc()->set_cluster_name("bar");
```

Comment:
How is this used?

Reply:
Good catch, this is a stub from when I was using `EXPECT_CALL(cm_.async_client_manager_, factoryForGrpcService(ProtoEq(expected_grpc_service)...))` in all the test cases. I'll get rid of it here...